src/lintsampler/grid.py

```diff
@@ -0,0 +1,292 @@
+                    "Grid.__init__: Edges not monotically increasing."
+                )
+            if not np.all(np.isfinite(arr)):
+                raise ValueError(
+                    "Grid.__init__: Edges not finite-valued."
+                )
+
+        # geometry-related attrs
+        self.mins = np.array([arr[0] for arr in self.edgearrays])
+        self.maxs = np.array([arr[-1] for arr in self.edgearrays])
+        self.shape = tuple(d - 1 for d in self.edgeshape)
+
+        # density-related attrs (None because densities not yet evaluated)
+        self.reset_densities()
+
+    def reset_densities(self):
+        """Unset density flag and remove density-related attributes.
+        
+        """
+        self.densities_evaluated = False
+        self.vertex_densities = None
+        self.masses = None
+        self.total_mass = None
+
+    def evaluate(self, pdf, vectorizedpdf=False, pdf_args=(), pdf_kwargs={}):
+        """Evaluate the user-provided pdf on grid and set related attributes.
+        
+        Parameters
+        ----------
+        pdf : function
+            Probability density function to evaluate on grid. Function should
+            take coordinate vector (or batch of vectors if vectorized; see
+            `vectorizedpdf` parameter) and return (unnormalised) density (or
+            batch of densities). Additional arguments can be passed to the
+            function via `pdf_args` and `pdf_kwargs` parameters.
+        vectorizedpdf : bool, optional
+            if True, assumes that the pdf function is vectorized, i.e., it
+            accepts  (..., k)-shaped batches of coordinate vectors and returns
+            (...)-shaped batches of densities. If False, assumes that the pdf
+            function simply accepts (k,)-shaped coordinate vectors and returns
+            single densities. Default is False.
+        pdf_args : tuple, optional
+            Additional positional arguments to pass to pdf function; function
+            call is `pdf(position, *pdf_args, **pdf_kwargs)`. Default is empty
+            tuple (no additional positional arguments).
+        pdf_kwargs : dict, optional
+            Additional keyword arguments to pass to pdf function; function call
+            is `pdf(position, *pdf_args, **pdf_kwargs)`. Default is empty dict
+            (no additional keyword arguments).            
+        """
+        # number of vertices in grid
+        npts = np.prod(self.edgeshape)
+        
+        # create the flattened grid for evaluation in k>1 case
+        if self.dim > 1:
+            g = np.stack(np.meshgrid(*self.edgearrays, indexing='ij'), axis=-1)
+            edgegrid = g.reshape(npts, self.dim)
+        else:
+            edgegrid = self.edgearrays[0]
+
+        # evaluate PDF over edge grid (iterate over grid if not vectorized)
+        if vectorizedpdf:
+            f = pdf(edgegrid, *pdf_args, **pdf_kwargs)
+            densities = np.array(f, dtype=np.float64)         
```

Comment:
This is a bit out there, but what if someone wanted even _more_ precision than `np.float64` for sampling at very low probabilities? Or is that case just too extreme?

Reply:
Yeah I suppose I could simply not specify a dtype here and if the PDF returns a more precise dtype then that would continue through in the vectorized case. But how do I do it in the non-vectorized case? Here the zeros array will have a dtype and then the values I fill in will take that dtype too.

Reply:
Good question. I guess I was also thinking of the simpler case, where one could perhaps get away with `np.float32`. But that may just be over-optimisation. Let's leave it.